libcxx/include/__flat_map/flat_map.h

```diff
@@ -0,0 +1,1326 @@
+    auto __last = __self.__containers_.keys.end();
+    if (__it == __last || __self.__compare_(__key, *__it)) {
+      return std::make_pair(__it, __it);
+    }
+    return std::make_pair(__it, std::next(__it));
+  }
+
+  template <class _Self, class _Kp>
+  _LIBCPP_HIDE_FROM_ABI static auto __equal_range_impl(_Self&& __self, const _Kp& __key) {
+    auto [__key_first, __key_last] = __key_equal_range(__self, __key);
+
+    const auto __make_mapped_iter = [&](const auto& __key_iter) {
+      return __self.__containers_.values.begin() +
+             static_cast<ranges::range_difference_t<mapped_container_type>>(
+                 ranges::distance(__self.__containers_.keys.begin(), __key_iter));
+    };
+
+    using __iterator_type = ranges::iterator_t<decltype(__self)>;
+    return std::make_pair(__iterator_type(__key_first, __make_mapped_iter(__key_first)),
+                          __iterator_type(__key_last, __make_mapped_iter(__key_last)));
+  }
+
+  template <class _Res, class _Self, class _Kp>
+  _LIBCPP_HIDE_FROM_ABI static _Res __lower_bound(_Self&& __self, _Kp& __x) {
+    return __binary_search<_Res>(__self, ranges::lower_bound, __x);
+  }
+
+  template <class _Res, class _Self, class _Kp>
+  _LIBCPP_HIDE_FROM_ABI static _Res __upper_bound(_Self&& __self, _Kp& __x) {
+    return __binary_search<_Res>(__self, ranges::upper_bound, __x);
+  }
+
+  template <class _Res, class _Self, class _Fn, class _Kp>
+  _LIBCPP_HIDE_FROM_ABI static _Res __binary_search(_Self&& __self, _Fn __search_fn, _Kp& __x) {
+    auto __key_iter = __search_fn(__self.__containers_.keys, __x, __self.__compare_);
+    auto __mapped_iter =
+        __self.__containers_.values.begin() +
+        static_cast<ranges::range_difference_t<mapped_container_type>>(
+            ranges::distance(__self.__containers_.keys.begin(), __key_iter));
+
+    return _Res(std::move(__key_iter), std::move(__mapped_iter));
+  }
+
+  template <class _KeyArg, class... _MArgs>
+  _LIBCPP_HIDE_FROM_ABI pair<iterator, bool> __try_emplace(_KeyArg&& __key, _MArgs&&... __mapped_args) {
+    auto __key_it    = ranges::lower_bound(__containers_.keys, __key, __compare_);
+    auto __mapped_it = __containers_.values.begin() + ranges::distance(__containers_.keys.begin(), __key_it);
+
+    if (__key_it == __containers_.keys.end() || __compare_(__key, *__key_it)) {
+      return pair<iterator, bool>(
+          __try_emplace_exact_hint(
+              std::move(__key_it),
+              std::move(__mapped_it),
+              std::forward<_KeyArg>(__key),
+              std::forward<_MArgs>(__mapped_args)...),
+          true);
+    } else {
+      return pair<iterator, bool>(iterator(std::move(__key_it), std::move(__mapped_it)), false);
+    }
+  }
+
+  template <class _Kp>
+  _LIBCPP_HIDE_FROM_ABI bool __is_hint_correct(const_iterator __hint, _Kp&& __key) {
+    // todo
+    // note that we have very fragile std::prev in our library
```

Comment:
This is actually really interesting. Our acceptance of non-bidi iterators in `std::prev` dates back to e1cd11d80f5929e916de861f83d672ab72e29046, which seems like it was fixing a bug. Personally, I'd be tempted to call this "not a bug" and to require bidirectional iterators in `std::prev`, but the Standard is also unclear about the requirements of `std::prev`. It does say `template <class BidirectionalIterator>`, but I can't find any normative wording that ties that into an actual "Constraints" or "Mandates" kind of clause.

It looks like libstdc++ behaves like libc++ (since we seem to have fixed the "bug" after they discovered it), but MSVC requires Bidirectional Iterators in `std::prev`: https://godbolt.org/z/qafxEc779.

Either way, I think `std::prev(non_bidi)` should be diagnosed at compile-time since it's so easy: https://github.com/llvm/llvm-project/issues/109456.

------------------

We still have a problem that `flat_map::iterator` is not a `Cpp17RandomAccessIterator`, which leads to really surprising and borderline dangerous behavior. For example, `std::advance(flat_map.begin(), 10)` would advance the iterator one-by-one because it can't tell that it's random access. That's nonsensical.

I think we have to figure out whether we're allowed to use `using iterator_category = random_access_iterator_tag;` for `flat_map::iterator` despite what http://eel.is/c++draft/flat.map#overview-1 says:

> flat_map supports iterators that meet the Cpp17InputIterator requirements and model the [random_access_iterator](http://eel.is/c++draft/iterator.concept.random.access#concept:random_access_iterator) concept ([[iterator.concept.random.access]](http://eel.is/c++draft/iterator.concept.random.access))[.](http://eel.is/c++draft/flat.map#overview-1.sentence-2)

Trying my luck with @timsong-cpp. Maybe @frederick-vs-ja has some context too?

Reply:
The reference isn't a real reference so it can't be anything above _Cpp17InputIterator_.

What are you doing for `vector<bool>`'s iterator? That has the same problem. This is really a general issue - perhaps the old advance/next/prev can check the new concepts too?

Reply:
> This is really a general issue - perhaps the old advance/next/prev can check the new concepts too?

It's really a pity that [P2408R5](https://wg21.link/p2408r5) wasn't generalized to old iterator positioning functions and iterator-pair constructors of containers/container adaptors.

I _guess_ it's fine now to "speculatively" generalized P2408R5 to remaining iterator-related utilities. Or if this is not desired, we may introduce some internal mechanisms to special-casing these proxy iterators. Special-casing will be possibly necessary once `hive` gets standardized.

Reply:
@timsong-cpp for `vector<bool>::iterator` , we have `using iterator_category = random_access_iterator_tag`. All our legacy algorithms are dispatched using the `iterator_category` tag. All these algorithms don't care whether or not they satisfy `Cpp17RandomAccessIterator` requirements. We could do the "cheat" here again by making the `iterator` `using iterator_category = random_access_iterator_tag`. But we are not sure if this is conforming

Reply:
I'm tempted to use `random_access_iterator_tag` here with a comment. I think that's less broken than using `input_iterator_tag`. And if we realize that this is non-conforming, we should change both `flat_map::iterator` and `vector<bool>::iterator` consistently.

Reply:
changed the `iterator_category` to `random_access_iterator_tag`